mpqp/qasm/open_qasm_2_and_3.py

```diff
@@ -740,3 +760,333 @@ def remove_include(qasm_code: str) -> str:
+    ):
+        m = re.match(
+            r"\s*([\w\d_]+)(\[.*?\])?\s*=\s*measure\s*([\w\d_]+)(\[.*?\])?\s*", instr
+        )
+        if m:
+            c, nb_c, q, nb_q = m.groups()
+            if nb_c and nb_q:
+                instructions_code += f"measure {q}{nb_q} -> {c}{nb_c};\n"
+            else:
+                instructions_code += f"measure {q} -> {c};\n"
+    elif instr_name == "u3":
+        header_code += add_qe_lib()
+        instructions_code += re.sub(r"\bu3\b", "u", instr) + ";\n"
+    elif instr_name == "cp":
+        header_code += add_qe_lib()
+        instructions_code += re.sub(r"\bcp\b", "cu1", instr) + ";\n"
+    elif instr_name in std_gates_3_to_2_map:
+        converted_instr_name = std_gates_3_to_2_map[instr_name]
+        instructions_code += (
+            re.sub(r"\b" + instr_name + r"\b", converted_instr_name, instr) + ";\n"
+        )
+    elif instr_name in std_gates_3:
+        m = re.match(r"\s*(.*)", instr)
+        if m:
+            instructions_code += m.group(1) + ";\n"
+    elif instr_name == "gate":
+        m = re.match(
+            r"\s*gate\s+(\w+)\s*(\(([^)]*)\))?\s*([\w\s,]*)\s*{\s*([^}]*)\s*}",
+            instr,
+            re.DOTALL,
+        )
+        if m:
+            gate_name = m.group(1)
+            params = f"({m.group(3)})" if m.group(3) else ""
+            qubits = m.group(4).strip()
+            body = m.group(5)
+
+            defined_gates.add(gate_name)
+            g_string = f"gate {gate_name}{params} {qubits} {{\n"
+            g_instructions = filter(
+                lambda i: not re.fullmatch(r"\s*", i), body.split(";")
+            )
+            for instruction in g_instructions:
+                instruction = instruction.strip()
+                i_code, h_code = convert_instruction_3_to_2(
+                    instruction,
+                    included_instr,
+                    included_tree_current,
+                    defined_gates,
+                    path_to_main,
+                )
+                g_string += " " * 4 + i_code  # Add indentation to body instructions
+                header_code += h_code
+
+            # Finalize the gate and add it to instructions_code
+            instructions_code += g_string + "}\n"
+
+    elif re.match(r"\s*if\s*\(.*?\)\s*.+", instr):
+        m = re.match(r"(\s*if\s*\(.*?\))\s*(.+)", instr)
+        if m:
+            if_statement = m.group(1)
+            nested_instr = m.group(2)
+            i_code, h_code = convert_instruction_3_to_2(
+                nested_instr,
+                included_instr,
+                included_tree_current,
+                defined_gates,
+                path_to_main,
+            )
+            instructions_code += if_statement + " " + i_code
+            header_code += h_code
+    elif instr_name in {"reset", "barrier"}:
+        instructions_code += instr + ";\n"
+    else:
+        gate = instr.split()[0].split("(")[0]
+        if gate == "ctrl":
```

Comment:
The keyword `ctrl` is only known in OpenQASM2, so we cannot just put it like this I think, and I don't see direct way of overcoming this ... since creating new gates from existing by controlling is not available in OpenQASM2.
Do y'all have any idea ?

Reply:
removed